crmsh/sh.py

```diff
@@ -0,0 +1,321 @@
+                input=cmd.encode('utf-8'),
+                **kwargs,
+            )
+
+
+class AutoShell:
+    def __init__(self, local_shell: LocalShell, user_of_host: UserOfHost):
+        self.local_shell = local_shell
+        self.user_of_host = user_of_host
+
+    def subprocess_run_no_input(self, host: typing.Optional[str], user: typing.Optional[str], cmd: str, **kwargs):
+        assert 'input' not in kwargs and 'stdin' not in kwargs
+        if host is None or host == self.local_shell.hostname():
+            return subprocess.run(
+                ['/bin/sh'],
+                input=cmd.encode('utf-8'),
+                **kwargs,
+            )
+        else:
+            if user is None:
+                user = 'root'
+            local_user, remote_user = self.user_of_host.user_pair_for_ssh(host)
+            return self.local_shell.su_subprocess_run(
+                local_user,
+                'ssh {} {}@{} sudo -H -u {} /bin/sh'.format(constants.SSH_OPTION, remote_user, host, user),
+                input=cmd.encode('utf-8'),
+                **kwargs,
+            )
+
+    def get_raw_stdout_stderr_no_input(self, host, cmd) -> typing.Tuple[int, bytes, bytes]:
+        result = self.subprocess_run_no_input(
+            host, None, cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        return result.returncode, result.stdout, result.stderr
+
+    def get_stdout_stderr_no_input(self, host, cmd) -> typing.Tuple[int, str, str]:
+        rc, stdout, stderr = self.get_raw_stdout_stderr_no_input(host, cmd)
+        return rc, Utils.decode_str(stdout).strip(), Utils.decode_str(stderr).strip()
+
+    def get_stdout_or_raise_error(
+            self,
+            cmd: str,
+            host: typing.Optional[str] = None,
+            success_exit_status: typing.Optional[typing.Set[int]] = None,
+    ):
+        rc, stdout, stderr = self.get_raw_stdout_stderr_no_input(host, cmd)
+        to_raise = False
+        if success_exit_status is None:
+            if rc != 0:
+                to_raise = True
+        else:
+            if rc not in success_exit_status:
+                to_raise = True
+        if not to_raise:
+            return Utils.decode_str(stdout).strip()
+        else:
+            raise CommandFailure(cmd, host, None, Utils.decode_str(stderr).strip())
+
+
+class LocalOnlyAutoShell(AutoShell):
+    def __init__(self, local_shell: LocalShell):
+        super().__init__(local_shell, None)
+
+    def subprocess_run_no_input(self, host: str, user: typing.Optional[str], cmd: str, **kwargs):
+        assert 'input' not in kwargs and 'stdin' not in kwargs
+        assert host is None or host == self.local_shell.hostname()
+        if user is None:
+            user = 'root'
+        return self.local_shell.su_subprocess_run(user, cmd, **kwargs)
+
+
+def auto_shell():
+    return AutoShell(LocalShell(), user_of_host.instance())
```

Comment:
I see there are still some functions in utils.py that calling `subprocess.Popen` or `subprocess.call` like `ext_cmd` `ext_cmd_nosudo` `get_stdout` and `get_stdout_stderr`

How about merging them into this module?

Reply:
```
    def get_stdout_stderr_raw
    def get_stdout_stderr
    def get_rc_and_error
    def get_stdout_or_raise_error
    def subprocess_run_no_input
    def get_raw_stdout_stderr_no_input
    def get_stdout_stderr_no_input
```
Do the names look confusing? How about
```
    def get_stdout_stderr_raw
    def get_stdout_stderr
    def get_rc_and_error
    def get_stdout_or_raise_error
    def subprocess_run_without_input
    def get_stdout_stderr_raw_without_input
    def get_stdout_stderr_without_input
```

Reply:
I moved `get_stdout` and `get_stdout_stderr` to `crmsh.sh`.

For direct calls to `subprocess.Popen` or `subprocess.call`, most of them use some advanced features provided by `subprocess` module, for example, piping data between 2 subprocesses. There is no need to create wrapper for these usages, or the wrapper class will be as complex as `subprocess` module.

For `ext_cmd` and `ext_cmd_sudo`, it seem that they are related the config `core.user`. I did not spent much time to inspect it. So I'd like to keep them unchanged.